### PR TITLE
bugfix: fix incorrect getAnnotation about class and method

### LINKED ISSUE #2562: 期望@GlobalTransactional支持ElementType.TYPE
期望@GlobalTransactional支持ElementType.TYPE，将注解添加到class上就可以该class上所有方法加上分布式事务

### PR DESCRIPTION
**Pull request by a364176773**
### Ⅰ. Describe what this PR did
bug fix for getAnnotation
1.类上有注解,方法无 可正常运行
2.方法有注解,类上无 可正常运行
3.类上方法上都有,优先读取方法上注解 可正常运行
4.基类方法调用时,排除基类调用 可正常运行
5.类重写基类方法,比如toString后,重写的方法属于本类方法,因此会开启全局事务.
### Ⅱ. Does this pull request fix one issue?

fix #2562

### CODE REVIEW

**Comment by zjinlei** [comment]
add test case

**Comment by a364176773** [comment]
> add test case

ok

File: spring/src/main/java/io/seata/spring/annotation/GlobalTransactionalInterceptor.java [hunk 1]
```diff
@@ -166,8 +165,11 @@ public TransactionInfo getTransactionInfo() {
...
 
     private <T extends Annotation> T getAnnotation(Method method, Class<?> targetClass, Class<T> annotationClass) {
-        return method == null ? targetClass == null ? null : targetClass.getAnnotation(annotationClass)
-            : method.getAnnotation(annotationClass);
+        T annotation = null == method ? null : method.getAnnotation(annotationClass);
+        if (null == annotation) {
+            annotation = null == targetClass ? null : targetClass.getAnnotation(annotationClass);
+        }
+        return annotation;
     }
```
Comment (helloworlde) [inline spring/src/main/java/io/seata/spring/annotation/GlobalTransactionalInterceptor.java:173]: How about to use lambda for more readable like :

```java
        return Optional.ofNullable(method)
                       .map(m -> m.getAnnotation(annotationClass))
                       .orElse(Optional.ofNullable(targetClass)
                                       .map(t -> t.getAnnotation(annotationClass))
                                       .orElse(null));
```

Comment (a364176773) [reply spring/src/main/java/io/seata/spring/annotation/GlobalTransactionalInterceptor.java:173]: thx

**Comment by a364176773** [comment]
@zjinlei PTAL

File: spring/src/test/java/io/seata/spring/annotation/MethodDescTest.java [hunk 2]
```diff
@@ -37,6 +36,8 @@
     private static Class<?> targetClass = null;
     private static GlobalTransactional transactional = null;
 
+    private static final GlobalTransactionalInterceptor globalTransactionalInterceptor=new GlobalTransactionalInterceptor(null);
```
Comment (jsbxyyx) [inline spring/src/test/java/io/seata/spring/annotation/MethodDescTest.java:39]: The global variable declare in testGetAnnotation method?

Comment (a364176773) [reply spring/src/test/java/io/seata/spring/annotation/MethodDescTest.java:39]: ok

Comment (a364176773) [reply spring/src/test/java/io/seata/spring/annotation/MethodDescTest.java]: > The global variable declare in testGetAnnotation method?

PTAL @jsbxyyx

File: spring/src/test/java/io/seata/spring/annotation/MethodDescTest.java [hunk 3]
```diff
@@ -30,13 +33,28 @@
...
+    private static Class<?> targetClass = null;
     private static GlobalTransactional transactional = null;
 
     public MethodDescTest() throws NoSuchMethodException {
         method = MockBusiness.class.getDeclaredMethod("doBiz", String.class);
         transactional = method.getAnnotation(GlobalTransactional.class);
     }
 
+    @Test
+    public void testGetAnnotation() throws NoSuchMethodException {
```
Comment (zjinlei) [inline spring/src/test/java/io/seata/spring/annotation/MethodDescTest.java:45]: 1. method, class both has Annotation
2.a. only method has Annotation, class is not null
2.b. method has Annotation, class is null
3.a. only class has Annotation, method is not null
3.b. only class has Annotation, method is null
lack 2.a and 3.a

File: spring/src/test/java/io/seata/spring/annotation/MethodDescTest.java [hunk 4]
```diff
@@ -30,13 +33,28 @@
...
+        GlobalTransactionalInterceptor globalTransactionalInterceptor = new GlobalTransactionalInterceptor(null);
+        Method method = MockBusiness.class.getDeclaredMethod("doBiz", String.class);
+        targetClass = Mockito.mock(MockBusiness.class).getClass();
+        transactional = globalTransactionalInterceptor.getAnnotation(method, targetClass, GlobalTransactional.class);
+        Assertions.assertEquals(transactional.timeoutMills(), 300000);
+        method = null;
+        transactional = globalTransactionalInterceptor.getAnnotation(method, targetClass, GlobalTransactional.class);
```
Comment (zjinlei) [inline spring/src/test/java/io/seata/spring/annotation/MethodDescTest.java:52]: add Assertions.assertEquals(transactional.timeoutMills(), TransactionInfo.DEFAULT_TIME_OUT * 2);

**Comment by a364176773** [comment]
@zjinlei @slievrly PTAL

**Comment by slievrly** [comment]
toString，hashcode these methods also take distributed transactions?

**Comment by yougecn** [comment]
测试了一下，注解加在类上、方法上、类及方法上混用、嵌套调用都没问题

**Comment by a364176773** [comment]
> toString，hashcode these methods also take distributed transactions?

PTAL@slievrly

**Comment by jsbxyyx** [comment]
add hashCode and equals ... test case?

**Pull request merged**